### PR TITLE
[Safety] Remove C Style Casts

### PR DESCRIPTION
**Pull request by Mytherin**
This PR removes most usages of C-style casts when casting to derived classes, such as when casting from e.g. a `ParsedExpression` to a `FunctionExpression`, or a `LogicalOperator` to a `LogicalGet`, etc. Instead, we replace these casts with a templated `Cast` function that returns a reference to the derived member. For example, this:

```cpp
void MyFunction(const ParsedExpression &expr) {
	auto &function = (FunctionExpression &) expr;
	...
}
```

Now becomes this:

```cpp
void MyFunction(const ParsedExpression &expr) {
	auto &function = expr.Cast<FunctionExpression>();
	...
}
```

This has a number of advantages over C-style casts:

* The `Cast` function will check that the cast is to the correct type (e.g. for expressions by looking at the `expression_class` variable) - an InternalException is thrown when a cast to an incorrect type is performed
* The `Cast` function cannot remove const-ness of values by accident. If `expr` is a `const`, then the returned value is also `const`.
* The `Cast` function will only compile if called correctly, as opposed to C-style casts. For example, this will compile (and crash):

```cpp
void MyFunction(const ParsedExpression *expr) {
	auto &function = (FunctionExpression &) expr; // EEK: casting a pointer to a reference
	...
}
```

But this will result in a compilation error:

```cpp
void MyFunction(const ParsedExpression *expr) {
	auto &function = expr.Cast<FunctionExpression>(); // compile error: need to use -> instead
	...
}
```

### CODE REVIEW

**Comment by Tishj** [comment]
I had a hard time finding an implementation of this, so for anyone else looking:
```c++
public:
	template <class TARGET>
	TARGET &Cast() {
		D_ASSERT(dynamic_cast<TARGET *>(this));
		return (TARGET &)*this;
	}

	template <class TARGET>
	const TARGET &Cast() const {
		D_ASSERT(dynamic_cast<const TARGET *>(this));
		return (const TARGET &)*this;
	}
```

**Comment by Mytherin** [comment]
I thought about trying that but opted not to for various reasons - most `Cast` functions are slightly different, and they are small enough that I don't feel bad copying them around a few times.

**Comment by carlopi** [comment]
Nice!
Potentially some checks could be moved at compile time (so that IFF starting type is known, it will fail to compile right away, but I am not sure it's feasible).

In any case if type is deducible ahead of time, it will be already properly code-generated (with the checking elided and either the call or the throw known), so that is a big nice gain.

**Comment by Tishj** [comment]
The thread sanitizer error is caused by the change to the concurrentqueue.hpp header guard, probably need to add to `.sanitizer-thread-suppressions.txt` I didn't know what exactly needs to be added.

`concurrentqueue.h` does already have a header guard, so the only thing this is now protecting against double include is the single threaded case - in which case hitting the thread sanitizer makes no sense?

File: benchmark/interpreted_benchmark.cpp [hunk 1]
```diff
@@ -382,7 +382,7 @@ unique_ptr<BenchmarkState> InterpretedBenchmark::Initialize(BenchmarkConfigurati
 		state->con.Query("PRAGMA enable_profiling");
 		state->con.Query("PRAGMA profiling_mode='detailed'");
 	}
-	return state;
+	return std::move(state);
```
Comment (Tishj) [inline benchmark/interpreted_benchmark.cpp:385]: Looks like you ran into #6964, I can't seem to reproduce this.
There are quite a lot of places we rely on this behavior, no?

**Pull request merged**